### ISSUE TITLE
Running multiple findUnique's in parallel causes both to return null

### ISSUE THREAD

**Issue by DustinJSilk** [prisma/prisma#4438]
## Bug description

When running two of the same findUnique queries in parallel, both return null where they should return the same document.

## How to reproduce

1. Create a new Prisma project with a PostgreSQL database.
2. Add a table with a @@unique attribute

```prisma
datasource db {
  provider = "postgresql"
  url      = "URL"
}

generator client {
  provider = "prisma-client-js"
}

model Property {
  id           String         @id @default(cuid())
  Availability Availability[]
}

model Availability {
  date          DateTime
  property      Property              @relation(fields: [propertyId], references: [id])
  propertyId    String
  status        AvailabilityStatus    @default(AVAILABLE)

  @@unique([propertyId, date, status])
}

enum AvailabilityStatus {
  AVAILABLE
}
```

3. Create a new set of documents in the table
```JavaScript
const property = await prisma.property.create({});

const targetDate = new Date('2020-12-03T00:00:00.000Z');

await prisma.availability.create({
    data: {
      property: { connect: {id: property.id }},
      status: AvailabilityStatus.AVAILABLE,
      date: targetDate,
    },
  });
```

4. Query the  document twice in parallel

```JavaScript
const firstUnique = prisma.availability.findUnique({
    where: {
      propertyId_date_status: {
        propertyId: property.id,
        status: AvailabilityStatus.AVAILABLE,
        date: targetDate,
      }
    },
  });

  const secondUnique = prisma.availability.findUnique({
    where: {
      propertyId_date_status: {
        propertyId: property.id,
        status: AvailabilityStatus.AVAILABLE,
        date: targetDate,
      }
    },
  });

  const res = await Promise.all([firstUnique, secondUnique])

  console.log(res); <-- [null, null]
```

## Expected behavior

Expected to return the same document twice - a warning would also be super helpful if this happens.

## Environment & setup

- OS: Mac
- Database: PostgreSQL
- Node.js version: 12.18.0
- Prisma version: 2.12.1

```
@prisma/cli          : 2.12.1
@prisma/client       : 2.12.1
Current platform     : darwin
Query Engine         : query-engine cf0680a1bfe8d5e743dc659cc7f08009f9587d58 (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli cf0680a1bfe8d5e743dc659cc7f08009f9587d58 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core cf0680a1bfe8d5e743dc659cc7f08009f9587d58 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt cf0680a1bfe8d5e743dc659cc7f08009f9587d58 (at node_modules/@prisma/engines/prisma-fmt-darwin)
Studio               : 0.322.0
```

**Comment by pantharshit00** [comment]
I can confirm this bug. Looks like multi column unique index is not working with the batch protocol. I can also reproduce this in the older version so this is not a regression but still this is a big bug. 

I have posted the reproduction here in simple script so that we can easily share it in the team: https://github.com/harshit-test-org/prisma-issue-4438

**Comment by Sytten** [comment]
This confirms that the batch protocol should have a way to be disabled for users that know what they are doing and don't want to use it.

**Comment by nrxus** [comment]
We just run into this behavior and it was *incredibly surprising*. The most disappointing part is how how silent this error is. There is no warning in the prisma logs, nothing documented about using composite keys with findUnique not working when doing things in parallel.  While I understand that this bug may be hard to resolve, at the very least some loud logging or just a panic so it can be caught quickly would be the least I would expect from a tool that should be prod-ready.

**Comment by Weakky** [comment]
Hey folks 👋,

We're no longer able to repro that bug in `2.15.0`. I'll therefore close the issue now. Feel free to chime-in/re-open if it persists.

@nrxus If you are already using `prisma@2.15.0`, please open another issue and describe your problem. You might be experiencing a different issue.

Thanks 🙏

**Comment by nrxus** [comment]
I do see it fixed in 2.15! Thank you so much!

I should have tried updating my prisma before commenting, I was on 2.14 and since I didn't see anything in the changelog for 2.15 about this I assumed it was still an issue 🤦🏽

**Comment by thomasjiangcy** [comment]
Currently using `4.10.1` and experience the same issue. Sample code:

```typescript
await Promise.all(users.map(user => await db.primary.profile.findUnique({
  where: {
    id: user.profileId
  }
})))
// Returns [null, null] assuming i have 2 users.
```

However it works when I use `.findUniqueOrThrow()` or `.findFirst()`

**Comment by janpio** [comment]
Please open a new issue and provide more information, and optimally a fully reproduction @thomasjiangcy. We can't debug this in a closed issue. Thanks.
